### PR TITLE
Add implementation

### PR DESCRIPTION
**Pull request by klzns**
#### What does this PR do? \*

Add implementation of a Challenge that checks if the user has the condition rules specified in the Trade Policy.

Important for B2B scenarios, where a Trade Policy is closed.

#### How to test it? \*

1. Open https://breno3--storecomponents.myvtex.com/?sc=3
2. You will be redirected to `/foobar`
3. Login using the user `breno@mailinator.com` `Abcd1234`
4. Go to the home
5. Now, you will not be redirected

Why is that, the Trade Policy 3 have the following setting:

image

This tells the backend to check if the logged in user has the propery `approved` as `true` in its MasterData data.

#### Related to / Depends on \*

> **Depends on https://github.com/vtex-apps/store-graphql/pull/423**

### CODE REVIEW

File: react/ChallengeSalesChannelCondition.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import React, { FC, useEffect, Fragment } from 'react'
+import { useRuntime, SessionUnauthorized } from 'vtex.render-runtime'
+import { useQuery } from 'react-apollo'
+import checkProfileAllowedQuery from './graphql/checkProfileAllowed.graphql'
+import withSession, { InjectedWithSessionProps } from './components/withSession'
+
+type ContentVisibility = 'visible' | 'hidden'
+
+const useRedirectForbidden = (
```
Comment (lbebber) [inline react/ChallengeSalesChannelCondition.tsx]:
```suggestion
const useRedirectIfForbidden = (
```

File: react/components/withSession.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      return null
+    }
+
+    return <WrappedComponent {...(props as P)} session={session} />
+  }
+
+  return WrappedComponentWithSession
+}
+
+export default withSession
```
Comment (lbebber) [inline react/components/withSession.tsx]: q tal `useSession` ao inves de um HOC?

File: react/components/withSession.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  session: Session | SessionUnauthorized
+}
+
+const withSession = <P extends object>(
+  WrappedComponent: ComponentType<P>
+): FC<P & WithSession> => {
+  const WrappedComponentWithSession: FC = (props: Record<string, unknown>) => {
+    const [session, setSession] = useState<Session | SessionUnauthorized>()
+
+    const sessionPromise =
```
Comment (lbebber) [inline react/components/withSession.tsx]: seria bom se isso viesse do render-runtime numa api melhor :/

File: react/ChallengeSalesChannelCondition.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    ssr: false,
+    skip: unauthorized,
+  })
+
+  const isAuthenticated =
+    unauthorized === true
+      ? false
+      : !loading && data && !error
+      ? !!data.checkProfileAllowed.allowed
+      : null
```
Comment (iaronaraujo) [inline react/ChallengeSalesChannelCondition.tsx]: This is a bit confusing, can you break this into normal `if`s?

**Review by iaronaraujo — CHANGES_REQUESTED** [review]

File: react/ChallengeTradePolicyCondition.tsx [hunk 5]
```diff
@@ -36,18 +59,16 @@ const useProfileAllowed = (skip: boolean) => {
...
   children,
 }) => {
-  const unauthorized =
-    session && (session as SessionUnauthorized).type === 'Unauthorized'
-  const profileAllowed = useProfileAllowed(unauthorized)
+  const isAuthorized = useSessionAuthorization()
+  const profileAllowed = useProfileAllowed(!isAuthorized)
 
-  const isAuthenticated = unauthorized === true ? false : profileAllowed
+  const isAuthenticated = isAuthorized === false ? false : profileAllowed
```
Comment (iaronaraujo) [inline react/ChallengeTradePolicyCondition.tsx:71]: Is this the same as `const isAuthenticated = isAuthorized || profileAllowed`?

File: react/ChallengeTradePolicyCondition.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    sessionPromise.then(sessionResponse => {
+      setSession(sessionResponse.response)
+    })
+  }, [sessionPromise])
+
+  if (session === undefined) {
+    return null
+  }
+
+  return session && (session as SessionUnauthorized).type === 'Unauthorized'
```
Comment (Klynger) [inline react/ChallengeTradePolicyCondition.tsx]:
```suggestion
  return session && (session as SessionUnauthorized).type.toLowerCase() === 'unauthorized'
```

**Review by iaronaraujo — DISMISSED** [review]

File: store/interfaces.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "challenge.tradePolicyCondition": {
```
Comment (Klynger) [inline store/interfaces.json]: Why not kebab?

File: react/ChallengeTradePolicyCondition.tsx (hunk 5 above)
Comment (klzns) [reply react/ChallengeTradePolicyCondition.tsx:71]: `isAuthorized` could be `null`, so no.

**Review by Klynger — APPROVED** [review]

**Pull request merged**